listenbrainz/webserver/static/js/src/LastFMImporter.tsx

```diff
@@ -60,14 +225,150 @@ export default class LastFmImporter extends React.Component<
     });
   };
 
-  setClose = () => {
-    this.setState({ canClose: this.importer.canClose });
+  setClose = (canClose: boolean) => {
+    this.setState({ canClose });
+  };
+
+  updateMessage = (msg: React.ReactElement) => {
+    this.setState({ msg });
   };
 
-  updateMessage = () => {
-    this.setState({ msg: this.importer.msg });
+  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    this.setState({ lastfmUsername: event.target.value });
+  };
+
+  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    const { lastfmUsername } = this.state;
+    this.toggleModal();
+    event.preventDefault();
+    this.startImport();
   };
 
+  async submitPage(payload: Array<Listen>) {
+    const delay = this.getRateLimitDelay();
+    // Halt execution for some time
+    await new Promise((resolve) => {
+      setTimeout(resolve, delay);
+    });
+
+    const response = await this.APIService.submitListens(
+      this.userToken,
+      "import",
+      payload
+    );
+    this.updateRateLimitParameters(response);
+  }
+
+  async startImport() {
+    this.updateModalAction(<p>Your import from Last.fm is starting!</p>, false);
+    this.latestImportTime = await this.APIService.getLatestImport(
+      this.userName
+    );
+    this.incrementalImport = this.latestImportTime > 0;
+    if (!this.incrementalImport)
+      this.playCount = await this.getTotalNumberOfScrobbles();
+    this.totalPages = await this.getNumberOfPages();
+    this.page = this.totalPages; // Start from the last page so that oldest scrobbles are imported first
+
+    while (this.page > 0) {
+      // Fixing no-await-in-loop will require significant changes to the code, ignoring for now
+      const payload = await this.getPage(this.page); // eslint-disable-line
+      if (payload) {
+        // Submit only if response is valid
+        this.submitPage(payload);
+      }
+
+      this.page -= 1;
+      this.numCompleted += 1;
+
+      // Update message
+      const msg = (
+        <p>
+          <FontAwesomeIcon icon={faSpinner as IconProp} spin /> Sending page{" "}
+          {this.numCompleted} of {this.totalPages} to ListenBrainz <br />
+          <span style={{ fontSize: `${8}pt` }}>
+            {this.incrementalImport && (
+              <span>
+                Note: This import will stop at the starting point of your last
+                import. :)
+                <br />
+              </span>
+            )}
+            <span>Please don&apos;t close this page while this is running</span>
+          </span>
+        </p>
+      );
+      this.updateMessage(msg);
```

Comment:
As this function doesn't do anything much now, we can just replace it with `this.setState({ msg })`